### ISSUE TITLE
Remove ensure_freshness for faster python?

### ISSUE THREAD

**Issue by pmichel31415** [clab/dynet#377]
Now that the c++ side checks for c++ expressions I think it's possible to remove most instances of `ensure_freshness` in the cython code because python will catch the c++ exception.

That would probably make expression creation faster in python.

**Comment by yoavg** [comment]
I did not follow the c++ changes closely, but if it already "ensures_frehsness" on the c++ side, then sure!
[though if you get "white" in `cython -a` for ensure freshness, then its likely will be a minimal gain. still worth doing if you have the bandwidth]

**Comment by pmichel31415** [comment]
Yeah I brought this up after seeing this in `cython -a`

![screenshot from 2017-03-13 11-51-53](https://cloud.githubusercontent.com/assets/10391785/23862528/9ac64558-07e3-11e7-818f-46f46bc965de.png)

![screenshot from 2017-03-13 11-47-51](https://cloud.githubusercontent.com/assets/10391785/23862359/297ab096-07e3-11e7-93c7-f7d70bacb330.png)

It might not bring a significant improvement but at least it'll make the code less verbose.

I will see if I can do this next week

**Comment by yoavg** [comment]
Yeah, I see no reason not to remove it, its better not to have code duplication across languages this way.

In terms of efficiency we'd probably not gain much though:
![image](https://cloud.githubusercontent.com/assets/464308/24045019/5b63096e-0b26-11e7-9e01-95bf05f438d0.png)
(the ensure_freshness `if` *is* surprisingly yellow, though :( )

**Comment by pmichel31415** [comment]
Good catch, good to know that `if` is so costly in cython.

**Comment by yoavg** [comment]
Its not the "if" that is costly, it what it operates on. Conditions on
primitive types are c-speed. You can click the beginning of the line in the
.html file to see the generated code.

**Comment by pmichel31415** [comment]
Ah got you, thanks for the explanation

**Comment by neubig** [comment]
Do you think this is worth doing? Or should we close the issue?

**Comment by pmichel31415** [comment]
Hmm so I still think this is worth doing but ofr another issue : in c++ we don't check whether expressions are stale when building operations (what `ensure_freshness` does) so say, if you do `x+y` and `y` is from a previous computation graph then you'll get a seg fault when running the forward pass.

TL;DR : we need to implement ensure_freshness in c++

Maybe this should be addressed in a separate issue, let me know.

**Comment by yoavg** [comment]
Oh, I though the c++ version was made safe in a similar way. As long there are no guards at the c++ level, I would like to keep the cython guards intact.

**Comment by pmichel31415** [comment]
I definitely agree, I opened a different issue for this ( #527 ), I'll close this one for now.
